flow/layers/image_filter_layer.cc

```diff
@@ -16,39 +18,66 @@ void ImageFilterLayer::Preroll(PrerollContext* context,
   Layer::AutoPrerollSaveLayerState save =
       Layer::AutoPrerollSaveLayerState::Create(context);
 
-  child_paint_bounds_ = SkRect::MakeEmpty();
-  PrerollChildren(context, matrix, &child_paint_bounds_);
+  SkRect child_bounds = SkRect::MakeEmpty();
+  PrerollChildren(context, matrix, &child_bounds);
   if (filter_) {
-    const SkIRect filter_input_bounds = child_paint_bounds_.roundOut();
+    const SkIRect filter_input_bounds = child_bounds.roundOut();
     SkIRect filter_output_bounds =
         filter_->filterBounds(filter_input_bounds, SkMatrix::I(),
                               SkImageFilter::kForward_MapDirection);
-    set_paint_bounds(SkRect::Make(filter_output_bounds));
-  } else {
-    set_paint_bounds(child_paint_bounds_);
+    child_bounds = SkRect::Make(filter_output_bounds);
   }
+  set_paint_bounds(child_bounds);
 
-  TryToPrepareRasterCache(context, this, matrix);
+  transformed_filter_ = nullptr;
+  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
+    // We have rendered this same ImageFilterLayer object enough
+    // times to consider its properties and children to be stable
+    // from frame to frame so we try to cache the layer itself
+    // for maximum performance.
+    TryToPrepareRasterCache(context, this, matrix);
+  } else if ((transformed_filter_ = filter_->makeWithLocalMatrix(matrix))) {
```

Comment:
Nit: it's probably very easy for someone to think `transformed_filter_ = filter_->makeWithLocalMatrix(matrix)` is a typo for `transformed_filter_ == filter_->makeWithLocalMatrix(matrix)`. A clearer approach might be
```
if (render_count >= ...) {
  // ...
  TryToPrepareRasterCache(context, this, matrix);
  return;
}

transformed_filter_ = filter_->makeWithLocalMatrix(matrix);
if (transformed_filter_) {
  ...
}
```

Reply:
Yes, I was going to do something similar and something inside me said "check it out, we could do it all in one line" and I didn't think about the maintenance issue.

I rewrote it similarly - I used an else instead of a return in case someone wants to later add more logic to Preroll after this section and doesn't spot the early return. An early return would be good if that block represented an action that would prevent any further logic, such as detecting an error condition, but in the case above, it is used to avoid the children caching case. Using an else clause keeps it self contained better - it's the "try the layer, or try the children as a backup" block chained into one overall statement.

Reply:
To be clear - I moved the assignment inside the body of the else clause and made it separate from the if statement that tested its return value:

```
  } else {
    transformed_filter_ = filter_->makeWithLocalMatrix(matrix);
    if (transformed_filter_) {
```
I think I'll add a comment about not being able to filter the children if the supplied ImageFilter can't "makeWithLocalMatrix"...

Reply:
There was also an implicit logic flaw fixed in the rewrite. If the filter transform combination fails, we should still increment the render count which that last version wasn't doing. The new version increments the render count regardless of the outcome of combining the filter and transform.

Reply:
Changes pushed in latest commit.